grails-app/services/org/pih/warehouse/inventory/CycleCountImportService.groovy

```diff
@@ -0,0 +1,73 @@
+        // Firstly clear the command's errors, as it might contain some additional errors related to the file itself
+        // that we do not care about while validating the data, and we don't them to be visible for a user
+        command.clearErrors()
+        // Store items that should be ignored for an import - for some cases we want not only to display an error, but also to ignore such row
+        List<Map> itemsToRemove = []
+        command.data.eachWithIndex { row, index ->
+            CycleCountRequest cycleCountRequest = CycleCount.read(row.cycleCountId)?.cycleCountRequest
+            // If a user provides an invalid cycle count id, skip that row and do not look for other errors, as they do not matter at this point
+            if (!cycleCountRequest) {
+                command.errors.reject("Row ${index + 1}: This row is skipped. Cycle count with given cycle count id ${row.cycleCountId} does not exist")
+                itemsToRemove.add(row)
+                // as explained above - since a cycle count is not found, there is no point to validate the row itself
+                return
+            }
+            CycleCountItem cycleCountItem = CycleCountItem.read(row.cycleCountItemId)
+            // If a cycle count item is not found, but user provided an id, do not treat such row as a custom row, but throw an error and ignore the row.
+            if (!cycleCountItem && row.cycleCountItemId) {
+                command.errors.reject("Row ${index + 1}: This row is skipped. Cycle count item with given cycle count item id ${row.cycleCountItemId} does not exist")
+                itemsToRemove.add(row)
+            }
+            // Ignore the row if product code is blank
+            if (!row.productCode) {
+                command.errors.reject("Row ${index + 1}: This row was ignored in the import because product code was blank - you must use a product associated with this cycle count")
+                itemsToRemove.add(row)
+            }
+            // Ignore the row is product code is different than the product code that cycle count is associated with
+            if (row.productCode != cycleCountRequest.product.productCode) {
+                command.errors.reject("Row ${index + 1}: This row is skipped, because you used an invalid product - you must use a product associated with this cycle count")
+                itemsToRemove.add(row)
+            }
+            // If bin location is not null, but an object with id and name properties, it means it is not a default bin
+            // but user provided an invalid location id
+            if (row.binLocation != null && !row.binLocation.id) {
+                command.errors.reject("Row ${index + 1}: Provided bin location does not exist in the system")
+            }
+            // Validation cases for an existing item (update)
+            // We mustn't update bin location, lot number and expiration date via import for an existing item
```

Comment:
it's a pretty rigid solution. I bet there will be users who will try to fix stock being in the incorrect bin or having the wrong lot by simply changing the bin or lot number for a row (that's how inventory import works). Since we have CCI id in the row, it does make sense that we need to fail here, it's just unfortunate because I know users will be trying to do this